Sources/SemanticIndex/UpdateIndexStoreTaskDescription.swift

```diff
@@ -353,33 +443,100 @@ package struct UpdateIndexStoreTaskDescription: IndexTaskDescription {
+    buildSettings.compilerArguments.removeSubrange(indexUnitOutputPathIndex...(indexUnitOutputPathIndex + 1))
+    if buildSettings.compilerArguments.contains("-index-unit-output-path") {
+      logger.error("Build settings contained two -index-unit-output-path arguments")
+      BuildSettingsLogger.log(settings: buildSettings, for: uri)
+    }
+    return indexUnitOutputPath
+  }
+
   private func updateIndexStore(
-    forSwiftFile uri: DocumentURI,
+    forSwiftFilesInPartition partition: UpdateIndexStorePartition,
     buildSettings: FileBuildSettings,
     toolchain: Toolchain
   ) async throws {
     guard let swiftc = toolchain.swiftc else {
       logger.error(
-        "Not updating index store for \(uri.forLogging) because toolchain \(toolchain.identifier) does not contain a Swift compiler"
+        "Not updating index store for \(partition.files) because toolchain \(toolchain.identifier) does not contain a Swift compiler"
       )
       return
     }
 
     var args =
       try [swiftc.filePath] + buildSettings.compilerArguments + [
         "-index-file",
-        "-index-file-path", uri.pseudoPath,
         // batch mode is not compatible with -index-file
         "-disable-batch-mode",
       ]
-    args = try await addOrReplaceIndexStorePath(in: args, for: uri)
+    args = try await addOrReplaceIndexStorePath(in: args, for: partition.files.map(\.mainFile))
+
+    switch partition {
+    case .multipleFiles(let filesAndOutputPaths, let buildSettings):
+      if await !toolchain.canIndexMultipleSwiftFilesInSingleInvocation {
+        // We should never get here because we shouldn't create `multipleFiles` batches if the toolchain doesn't support
+        // indexing multiple files in a single compiler invocation.
+        logger.fault("Cannot index multiple files in a single compiler invocation.")
+      }
 
-    try await runIndexingProcess(
-      indexFile: uri,
-      buildSettings: buildSettings,
-      processArguments: args,
-      workingDirectory: buildSettings.workingDirectory.map(AbsolutePath.init(validating:))
-    )
+      struct OutputFileMapEntry: Encodable {
+        let indexUnitOutputPath: String
+
+        private enum CodingKeys: String, CodingKey {
+          case indexUnitOutputPath = "index-unit-output-path"
+        }
+      }
+      var outputFileMap: [String: OutputFileMapEntry] = [:]
+      for (fileInfo, outputPath) in filesAndOutputPaths {
+        guard let filePath = try? fileInfo.mainFile.fileURL?.filePath else {
+          logger.error("Failed to determine file path of file to index \(fileInfo.mainFile.forLogging)")
+          continue
+        }
+        outputFileMap[filePath] = .init(indexUnitOutputPath: outputPath)
+      }
+      let tempFileUri = FileManager.default.temporaryDirectory
+        .appending(component: "sourcekit-lsp-output-file-map-\(UUID().uuidString).json")
+      try JSONEncoder().encode(outputFileMap).write(to: tempFileUri)
```

Comment:
Do we ever delete this?

Reply:
Ooops, must have lost the `defer` block to delete this at some point.